stdlib/public/Cxx/CxxOptional.swift

```diff
@@ -0,0 +1,45 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+public protocol CxxOptional<Wrapped> {
+  associatedtype Wrapped
+
+  func __convertToBool() -> Bool
+
+  var pointee: Wrapped { get }
+}
+
+extension CxxOptional {
+  @inlinable
+  public var hasValue: Bool {
+    get {
+      return __convertToBool()
+    }
+  }
+
+  @inlinable
+  public var value: Wrapped {
+    get {
+      return pointee
+    }
+  }
+}
+
+extension Optional {
+  public init<O: CxxOptional>(fromCxx value: O) where O.Wrapped == Wrapped {
```

Comment:
This probably ought to be marked `@inlinable` to allow specializations. Additionally, it is good style to use the more readable generic declaration syntax -- it will end up in API docs and it will make it much easier for readers to understand the signature:

```suggestion
  @inlinable
  public init(fromCxx value: some CxxOptional<Wrapped>) {
```

Reply:
Thanks! I adjusted the signature.